FHEM/00_SIGNALduino.pm

```diff
@@ -2221,95 +2223,122 @@ sub SIGNALduino_Parse_MU($$$$@)
 			my $pstr="";
+			my $zeroStr ="";
+			my $oneStr ="";
+			my $floatStr ="";
+			my $signal_regex;
+			my $start_regex;
+			
 			$valid = $valid && ($pstr=SIGNALduino_PatternExists($hash,\@{$ProtocolListSIGNALduino{$id}{one}},\%patternList,\$rawData)) >=0;
 			Debug "Found matched one" if ($debug && $valid);
-			my $oneStr=$pstr if ($valid);
-			$patternLookupHash{$pstr}="1" if ($valid); ## Append one to our lookuptable
+			$oneStr=$pstr if ($valid);
+			$patternLookupHash{$pstr}="1" if ($valid); 		## Append one to our lookuptable
 			Debug "added $pstr " if ($debug && $valid);
-
-			my $zeroStr ="";
+			
 			if (scalar @{$ProtocolListSIGNALduino{$id}{zero}} >0) {
 				$valid = $valid && ($pstr=SIGNALduino_PatternExists($hash,\@{$ProtocolListSIGNALduino{$id}{zero}},\%patternList,\$rawData)) >=0;
 				Debug "Found matched zero" if ($debug && $valid);
-				$zeroStr=$pstr if ($valid);
-				$patternLookupHash{$pstr}="0" if ($valid); ## Append zero to our lookuptable
+				$zeroStr='|' . $pstr if ($valid);
+				$patternLookupHash{$pstr}="0" if ($valid);	## Append zero to our lookuptable
 				Debug "added $pstr " if ($debug && $valid);
 			}
 
-			if (defined($ProtocolListSIGNALduino{$id}{float}))
-			{
+			if (defined($ProtocolListSIGNALduino{$id}{float})) {
 				$valid = $valid && ($pstr=SIGNALduino_PatternExists($hash,\@{$ProtocolListSIGNALduino{$id}{float}},\%patternList,\$rawData)) >=0;
 				Debug "Found matched float" if ($debug && $valid);
-				$patternLookupHash{$pstr}="F" if ($valid); ## Append float to our lookuptable
+				$floatStr='|' . $pstr if ($valid);
+				$patternLookupHash{$pstr}="F" if ($valid);	## Append float to our lookuptable
 				Debug "added $pstr " if ($debug && $valid);
 			}
 
-			next if (!$valid) ;
-			#Debug "Pattern Lookup Table".Dumper(%patternLookupHash);
-			## Check somethin else
-
-		
-			#Anything seems to be valid, we can start decoding this.			
-
-			SIGNALduino_Log3 $name, 4, "$name: Fingerprint for MU Protocol id $id -> $ProtocolListSIGNALduino{$id}{name} matches, trying to demodulate"  if ($valid);
-			my $signal_width= @{$ProtocolListSIGNALduino{$id}{one}};
-			#Debug $signal_width;
+			next if (!$valid);
 			
-			my @bit_msg=();							# array to store decoded signal bits
+			SIGNALduino_Log3 $name, 4, "$name: Fingerprint for MU Protocol id $id -> $ProtocolListSIGNALduino{$id}{name} matches, trying to demodulate";
 			
-			my $message_start=0 ;
-			my @msgStartLst;
-			my $startStr="";
-			my $start_regex;
-			#my $oneStr=SIGNALduino_PatternExists($hash,\@{$ProtocolListSIGNALduino{$id}{one}},\%patternList,\$rawData);
-			#my $zeroStr=SIGNALduino_PatternExists($hash,\@{$ProtocolListSIGNALduino{$id}{zero}},\%patternList,\$rawData);
+			my $signal_width= @{$ProtocolListSIGNALduino{$id}{one}};
+			my $length_min;
+			if (defined($ProtocolListSIGNALduino{$id}{length_min})) {
+				$length_min = $ProtocolListSIGNALduino{$id}{length_min};
+			} else {
+				$length_min = SDUINO_PARSE_DEFAULT_LENGHT_MIN;
```

Comment:
Ich bin nicht sicher, ob es klug ist, die Zuweisung nur in Parse MU und bei jedem Durchlauf vorzunehmen. An dem Zustand ändert sich zur Laufzeit nichts.
Mir würde es besser gefallen, wenn wir Standards global setzen und es dann auch für alle Protokolle gilt.

Der passende Platz wäre vermutlich ungefähr hier: https://github.com/RFD-FHEM/RFFHEM/blob/23fda87658788ed975032a530a6282e840decacc/FHEM/00_SIGNALduino.pm#L2886

Das setzen von Standardwerten am besten in einer eigenen sub implementieren. Damit lässt sich diese dann einfacher und gezielter verifizieren.

Reply:
Die constant SDUINO_PARSE_DEFAULT_LENGHT_MIN ist am Anfang definiert und
$length_min = $ProtocolListSIGNALduino{$id}{length_min};
ist abhängig von der ID.
Möchstest Du die length_min in einem array $length_min[$id] speichern?

Reply:
Ich würde dazu tendieren den Hash mit Standards zu befüllen, wenn diese nicht angegeben wurden.

Also
$ProtocolListSIGNALduino{$id}{length_min} = SDUINO_PARSE_DEFAULT_LENGHT_MIN if (!exists($ProtocolListSIGNALduino{$id}{length_min});

Dann ist der Standadwert überall im code verfügbar und nicht nur an dieser Stelle.

Reply:
Ich habe das mit dem length_min erstmal in der Parse_MU gelassen, dies kann dann in einem weiteren Schritt in die Sub SIGNALduino_IdList() verlagert werden.
Evtl macht es Sinn dies auch bei den MS-Nachrichten zu machen.